controllers/healthz/health_reconciler.go

```diff
@@ -0,0 +1,172 @@
+// limitations under the License.
+
+package healthz
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/gardener/controller-manager-library/pkg/logger"
+	druidv1alpha1 "github.com/gardener/etcd-druid/api/v1alpha1"
+	"github.com/gardener/etcd-druid/pkg/client/kubernetes"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	// reconcilerName is the name of the controller
+	reconcilerName = "healthz_reconciler"
+	healthyLabel   = "healthy"
+)
+
+// Reconciler is health recocniler object for external service created on etcd
+type Reconciler struct {
+	clietnset kubernetes.Interface
+	client    client.Client
+}
+
+// NewHealthReconciler return the health reconciler object for external service created on etcd
+func NewHealthReconciler(cli kubernetes.Interface) reconcile.Reconciler {
+	return &Reconciler{
+		clietnset: cli,
+		client:    cli.Client(),
+	}
+}
+
+// Reconcile reconciles the etcd health.
+func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.TODO()
+
+	etcd := &druidv1alpha1.Etcd{}
+	if err := r.client.Get(ctx, req.NamespacedName, etcd); err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	logger.WithLogger(ctx, "reconciler", reconcilerName)
+	logger.Infof("Reconciling etcd for health: %s/%s", etcd.GetNamespace(), etcd.GetName())
+	if etcd.Status.ServiceName == nil {
+		logger.Infof("missing external service. Skipping reconciliation.")
+		return ctrl.Result{
+			RequeueAfter: time.Second * 5,
+		}, nil
+	}
+
+	// TODO: Discuss getBackupHealthOverExec vs getBackupHealthOverHTTP
+	healthy, err := r.getBackupHealthOverHTTP(ctx, etcd)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	svc := &corev1.Service{}
+	if err := r.client.Get(ctx, types.NamespacedName{Name: *etcd.Status.ServiceName, Namespace: etcd.Namespace}, svc); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if healthy {
+		svcCopy := svc.DeepCopy()
+		if svcCopy.Spec.Selector != nil {
+			if _, ok := svcCopy.Spec.Selector[healthyLabel]; ok {
+				// if present then only do the patch
+				delete(svcCopy.Spec.Selector, healthyLabel)
+
+				if err := r.client.Patch(ctx, svcCopy, client.MergeFrom(svc)); err != nil {
+					return ctrl.Result{}, err
+				}
+			}
+		}
+	} else {
+		svcCopy := svc.DeepCopy()
+		if svcCopy.Spec.Selector != nil {
+			if _, ok := svcCopy.Spec.Selector[healthyLabel]; !ok {
+				// if not present then only add label and patch.
+				svcCopy.Spec.Selector[healthyLabel] = "false"
```

Comment:
Adding the key/value of `healthy: "false"` to the service selector works but the choice of the label key and value could be confusing for people who do not know that the pods will never be labelled with this key/value and it is deliberately being added to avoid the service from pointing to the pods. Maybe something like `blocked` or `disabled` some even uglier and longer name might be clearer.